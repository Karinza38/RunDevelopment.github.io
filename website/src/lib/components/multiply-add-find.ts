import { lazyCall } from "../util";
import {
    bruteForceAllSolutions,
    Conversion,
    ConversionRange,
    Request,
} from "./multiply-add-constants";

interface Options {
    optimizeAdd?: boolean;
    fullShifts?: number;
}

interface BruteForceResult {
    conversion: Conversion | undefined;
    solutions: ConversionRange[];
}

async function bruteForceSolution(
    request: Request,
    options: Options,
    signal?: AbortSignal,
): Promise<BruteForceResult> {
    const pickAny = (range: ConversionRange): Conversion => {
        const { factor, shift } = range;
        const [addMin, addMax] = range.add;

        if (addMin === 0) {
            // 0 is the optimal add value, because it's free
            return { factor, add: 0, shift };
        }
        if (addMin === 1) {
            // +1 has a very cheap hardware implementation
            return { factor, add: 1, shift };
        }
        if (addMin <= factor && factor <= addMax) {
            // the compiler can optimize f*x+f to f*(x+1)
            return { factor, add: factor, shift };
        }
        if (addMax === 2 ** shift - 1) {
            return { factor, add: addMax, shift };
        }
        return { factor, add: addMin, shift };
    };

    const globalScope: Record<string, unknown> =
        typeof globalThis !== "undefined" ? globalThis : window;
    const bruteForceAll =
        "bruteForceAllSolutions" in globalScope
            ? (globalScope["bruteForceAllSolutions"] as typeof bruteForceAllSolutions)
            : bruteForceAllSolutions;

    if (options.fullShifts) {
        let conversion: Conversion | undefined;
        const solutions: ConversionRange[] = [];
        for await (const solution of bruteForceAll(
            request,
            { maxShiftAfterFirstSolution: options.fullShifts },
            signal,
        )) {
            conversion ??= pickAny(solution);
            solutions.push(solution);
        }
        return { conversion, solutions };
    }

    if (!options.optimizeAdd) {
        // any solution will do
        for await (const solution of bruteForceAll(request, {}, signal)) {
            return { conversion: pickAny(solution), solutions: [solution] };
        }
        return { conversion: undefined, solutions: [] };
    }

    let any: Conversion | undefined;
    let factorAdd: Conversion | undefined;
    const solutions: ConversionRange[] = [];
    for await (const solution of bruteForceAll(
        request,
        { maxShiftAfterFirstSolution: 6 },
        signal,
    )) {
        solutions.push(solution);
        any ??= pickAny(solution);
        if (solution.add[0] === 0) {
            return { conversion: pickAny(solution), solutions };
        }
        if (factorAdd === null) {
            if (solution.add[0] <= solution.factor && solution.factor <= solution.add[1]) {
                factorAdd = {
                    factor: solution.factor,
                    add: solution.factor,
                    shift: solution.shift,
                };
            }
        }
    }

    return { conversion: factorAdd ?? any, solutions };
}

function webWorkerBruteForce() {
    let lastAbort: AbortController | null = null;
    let lastPromise: Promise<unknown> = Promise.resolve();
    const onmessage = (e: MessageEvent) => {
        const { id, request, requirements } = e.data;
        const process = async () => {
            try {
                lastAbort?.abort("Took too long");
                await lastPromise;
                const abort = new AbortController();
                lastAbort = abort;

                const start = Date.now();
                const promise = bruteForceSolution(request, requirements, abort.signal);
                lastPromise = promise.catch(() => {});
                const result = await promise;
                const time = Date.now() - start;
                postMessage({ id, ...result, request, time });
            } catch (e) {
                postMessage({ id, error: String(e) });
            }
        };
        process().catch(console.error);
    };

    const sourceCode =
        bruteForceAllSolutions.toString() +
        bruteForceSolution.toString() +
        "\nlet lastAbort = null;" +
        "\nlet lastPromise = Promise.resolve();" +
        "\nonmessage = " +
        onmessage.toString();
    const worker = new Worker(
        URL.createObjectURL(new Blob([sourceCode], { type: "application/javascript" })),
    );

    let idCounter = 0;
    const listeners = new Map<number, (result: unknown) => void>();

    worker.onmessage = (e) => {
        const { id } = e.data;
        const listener = listeners.get(id);
        if (listener) {
            listener(e.data);
            listeners.delete(id);
        }
    };

    return (request: Request, requirements: Options) => {
        const id = idCounter++;
        worker.postMessage({ id, request, requirements });
        return new Promise<SearchResult>((resolve, reject) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            listeners.set(id, (result: any) => {
                if (result.error) {
                    reject(new Error(result.error));
                } else {
                    resolve({
                        conversion: result.conversion,
                        solutions: result.solutions,
                        time: result.time,
                        request: request,
                    });
                }
            });
        });
    };
}

export interface SearchResult {
    conversion: Conversion | undefined;
    solutions: ConversionRange[];
    request: Request;
    time: number;
}

export async function findConversion(
    request: Request,
    options: Options = {},
): Promise<SearchResult> {
    if (typeof Worker !== "undefined") {
        return lazyCall(webWorkerBruteForce)(request, options);
    }
    const start = Date.now();
    const result = await bruteForceSolution(request, options);
    const time = Date.now() - start;
    return { ...result, time, request };
}
