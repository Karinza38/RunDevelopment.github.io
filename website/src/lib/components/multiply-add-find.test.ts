import { findConversion } from "./multiply-add-find";

describe("findConversion unorm", () => {
    it("should find unorm conversions", async () => {
        const MAX_UNORM_BITS = 16;

        const results: string[] = [];

        const promises: Promise<void>[] = [];
        for (let from = 1; from <= MAX_UNORM_BITS; from++) {
            for (let to = 1; to <= MAX_UNORM_BITS; to++) {
                const key = `${String(from).padStart(2)} -> ${String(to).padEnd(2)}`;

                const inputRange = 2 ** from - 1;
                const outputRange = 2 ** to - 1;
                promises.push(
                    findConversion({ inputRange, D: inputRange, T: outputRange, R: "round" })
                        .then(
                            (res) => {
                                if (!res.conversion) {
                                    return `No conversion found`;
                                }
                                const { factor, add, shift } = res.conversion;
                                return `${factor} ${add} ${shift}`;
                            },
                            (err) => {
                                return `Error: ${err.message}`;
                            },
                        )
                        .then((res) => {
                            results.push(`${key}   ${res}`);
                        }),
                );
            }
        }
        await Promise.all(promises);

        expect(results).toMatchSnapshot();
    });
});
