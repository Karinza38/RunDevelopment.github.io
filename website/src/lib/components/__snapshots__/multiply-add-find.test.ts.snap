// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findConversion unorm should find unorm conversions 1`] = `
[
  " 1 -> 1    1 0 0",
  " 1 -> 2    3 0 0",
  " 1 -> 3    7 0 0",
  " 1 -> 4    15 0 0",
  " 1 -> 5    31 0 0",
  " 1 -> 6    63 0 0",
  " 1 -> 7    127 0 0",
  " 1 -> 8    255 0 0",
  " 1 -> 9    511 0 0",
  " 1 -> 10   1023 0 0",
  " 1 -> 11   2047 0 0",
  " 1 -> 12   4095 0 0",
  " 1 -> 13   8191 0 0",
  " 1 -> 14   16383 0 0",
  " 1 -> 15   32767 0 0",
  " 1 -> 16   65535 0 0",
  " 2 -> 1    1 0 1",
  " 2 -> 2    1 0 0",
  " 2 -> 3    5 0 1",
  " 2 -> 4    5 0 0",
  " 2 -> 5    21 0 1",
  " 2 -> 6    21 0 0",
  " 2 -> 7    85 0 1",
  " 2 -> 8    85 0 0",
  " 2 -> 9    341 0 1",
  " 2 -> 10   341 0 0",
  " 2 -> 11   1365 0 1",
  " 2 -> 12   1365 0 0",
  " 2 -> 13   5461 0 1",
  " 2 -> 14   5461 0 0",
  " 2 -> 15   21845 0 1",
  " 2 -> 16   21845 0 0",
  " 3 -> 1    1 0 2",
  " 3 -> 2    1 0 1",
  " 3 -> 3    1 0 0",
  " 3 -> 4    9 0 2",
  " 3 -> 5    9 0 1",
  " 3 -> 6    9 0 0",
  " 3 -> 7    73 0 2",
  " 3 -> 8    73 0 1",
  " 3 -> 9    73 0 0",
  " 3 -> 10   585 0 2",
  " 3 -> 11   585 0 1",
  " 3 -> 12   585 0 0",
  " 3 -> 13   4681 0 2",
  " 3 -> 14   4681 0 1",
  " 3 -> 15   4681 0 0",
  " 3 -> 16   37449 0 2",
  " 4 -> 1    1 0 3",
  " 4 -> 2    3 9 4",
  " 4 -> 3    1 0 1",
  " 4 -> 4    1 0 0",
  " 4 -> 5    17 0 3",
  " 4 -> 6    67 9 4",
  " 4 -> 7    17 0 1",
  " 4 -> 8    17 0 0",
  " 4 -> 9    273 0 3",
  " 4 -> 10   1091 9 4",
  " 4 -> 11   273 0 1",
  " 4 -> 12   273 0 0",
  " 4 -> 13   4369 0 3",
  " 4 -> 14   17475 9 4",
  " 4 -> 15   4369 0 1",
  " 4 -> 16   4369 0 0",
  " 5 -> 1    1 0 4",
  " 5 -> 2    13 50 7",
  " 5 -> 3    15 23 6",
  " 5 -> 4    1 0 1",
  " 5 -> 5    1 0 0",
  " 5 -> 6    33 0 4",
  " 5 -> 7    525 50 7",
  " 5 -> 8    527 23 6",
  " 5 -> 9    33 0 1",
  " 5 -> 10   33 0 0",
  " 5 -> 11   1057 0 4",
  " 5 -> 12   16909 50 7",
  " 5 -> 13   16911 23 6",
  " 5 -> 14   1057 0 1",
  " 5 -> 15   1057 0 0",
  " 5 -> 16   33825 0 4",
  " 6 -> 1    1 0 5",
  " 6 -> 2    3 33 6",
  " 6 -> 3    7 35 6",
  " 6 -> 4    61 121 8",
  " 6 -> 5    1 0 1",
  " 6 -> 6    1 0 0",
  " 6 -> 7    65 0 5",
  " 6 -> 8    259 33 6",
  " 6 -> 9    519 35 6",
  " 6 -> 10   4157 121 8",
  " 6 -> 11   65 0 1",
  " 6 -> 12   65 0 0",
  " 6 -> 13   4161 0 5",
  " 6 -> 14   16643 33 6",
  " 6 -> 15   33287 35 6",
  " 6 -> 16   266301 121 8",
  " 7 -> 1    1 0 6",
  " 7 -> 2    49 970 11",
  " 7 -> 3    57 454 10",
  " 7 -> 4    121 504 10",
  " 7 -> 5    125 252 9",
  " 7 -> 6    1 0 1",
  " 7 -> 7    1 0 0",
  " 7 -> 8    129 0 6",
  " 7 -> 9    8241 970 11",
  " 7 -> 10   8249 454 10",
  " 7 -> 11   16505 504 10",
  " 7 -> 12   16509 252 9",
  " 7 -> 13   129 0 1",
  " 7 -> 14   129 0 0",
  " 7 -> 15   16513 0 6",
  " 7 -> 16   1056817 970 11",
  " 8 -> 1    1 0 7",
  " 8 -> 2    3 129 8",
  " 8 -> 3    225 4060 13",
  " 8 -> 4    15 135 8",
  " 8 -> 5    249 1014 11",
  " 8 -> 6    253 505 10",
  " 8 -> 7    1 0 1",
  " 8 -> 8    1 0 0",
  " 8 -> 9    257 0 7",
  " 8 -> 10   1027 129 8",
  " 8 -> 11   65761 4060 13",
  " 8 -> 12   4111 135 8",
  " 8 -> 13   65785 1014 11",
  " 8 -> 14   65789 505 10",
  " 8 -> 15   257 0 1",
  " 8 -> 16   257 0 0",
  " 9 -> 1    1 0 8",
  " 9 -> 2    193 16170 15",
  " 9 -> 3    7 259 9",
  " 9 -> 4    241 3854 13",
  " 9 -> 5    497 4080 13",
  " 9 -> 6    505 2019 12",
  " 9 -> 7    509 1020 11",
  " 9 -> 8    1 0 1",
  " 9 -> 9    1 0 0",
  " 9 -> 10   513 0 8",
  " 9 -> 11   131265 16170 15",
  " 9 -> 12   4103 259 9",
  " 9 -> 13   131313 3854 13",
  " 9 -> 14   262641 4080 13",
  " 9 -> 15   262649 2019 12",
  " 9 -> 16   262653 1020 11",
  "10 -> 1    1 0 9",
  "10 -> 2    3 513 10",
  "10 -> 3    449 32310 16",
  "10 -> 4    961 32653 16",
  "10 -> 5    31 527 10",
  "10 -> 6    1009 8165 14",
  "10 -> 7    1017 4088 13",
  "10 -> 8    1021 2041 12",
  "10 -> 9    1 0 1",
  "10 -> 10   1 0 0",
  "10 -> 11   1025 0 9",
  "10 -> 12   4099 513 10",
  "10 -> 13   524737 32310 16",
  "10 -> 14   1049537 32653 16",
  "10 -> 15   32799 527 10",
  "10 -> 16   1049585 8165 14",
  "11 -> 1    1 0 10",
  "11 -> 2    769 261290 19",
  "11 -> 3    1793 261852 19",
  "11 -> 4    1921 130936 18",
  "11 -> 5    993 31774 16",
  "11 -> 6    2017 32736 16",
  "11 -> 7    2033 16362 15",
  "11 -> 8    2041 8182 14",
  "11 -> 9    2045 4092 13",
  "11 -> 10   1 0 1",
  "11 -> 11   1 0 0",
  "11 -> 12   2049 0 10",
  "11 -> 13   2097921 261290 19",
  "11 -> 14   4196097 261852 19",
  "11 -> 15   4196225 130936 18",
  "11 -> 16   2098145 31774 16",
  "12 -> 1    1 0 11",
  "12 -> 2    3 2049 12",
  "12 -> 3    7 2051 12",
  "12 -> 4    15 2055 12",
  "12 -> 5    3969 261996 19",
  "12 -> 6    63 2079 12",
  "12 -> 7    4065 65494 17",
  "12 -> 8    4081 32647 16",
  "12 -> 9    4089 16355 15",
  "12 -> 10   4093 8185 14",
  "12 -> 11   1 0 1",
  "12 -> 12   1 0 0",
  "12 -> 13   4097 0 11",
  "12 -> 14   16387 2049 12",
  "12 -> 15   32775 2051 12",
  "12 -> 16   65551 2055 12",
  "13 -> 1    1 0 12",
  "13 -> 2    3073 4190890 23",
  "13 -> 3    3585 2093494 22",
  "13 -> 4    3841 1044718 21",
  "13 -> 5    7937 1048230 21",
  "13 -> 6    4033 258110 19",
  "13 -> 7    8129 262080 19",
  "13 -> 8    8161 131036 18",
  "13 -> 9    8177 65520 17",
  "13 -> 10   8185 32760 16",
  "13 -> 11   8189 16380 15",
  "13 -> 12   1 0 1",
  "13 -> 13   1 0 0",
  "13 -> 14   8193 0 12",
  "13 -> 15   33557505 4190890 23",
  "13 -> 16   33558017 2093494 22",
  "14 -> 1    1 0 13",
  "14 -> 2    3 8193 14",
  "14 -> 3    14337 16774876 25",
  "14 -> 4    15361 8386765 24",
  "14 -> 5    15873 4193776 23",
  "14 -> 6    16129 2096685 22",
  "14 -> 7    127 8255 14",
  "14 -> 8    16321 524181 20",
  "14 -> 9    16353 262098 19",
  "14 -> 10   16369 131045 18",
  "14 -> 11   16377 65526 17",
  "14 -> 12   16381 32761 16",
  "14 -> 13   1 0 1",
  "14 -> 14   1 0 0",
  "14 -> 15   16385 0 13",
  "14 -> 16   65539 8193 14",
  "15 -> 1    1 0 14",
  "15 -> 2    12289 67095210 27",
  "15 -> 3    7 16387 15",
  "15 -> 4    30721 33552248 26",
  "15 -> 5    31 16399 15",
  "15 -> 6    32257 8386747 24",
  "15 -> 7    16257 2080894 22",
  "15 -> 8    32641 2097024 22",
  "15 -> 9    32705 1048347 21",
  "15 -> 10   32737 523791 20",
  "15 -> 11   32753 262122 19",
  "15 -> 12   32761 131043 18",
  "15 -> 13   32765 65532 17",
  "15 -> 14   1 0 1",
  "15 -> 15   1 0 0",
  "15 -> 16   32769 0 14",
  "16 -> 1    1 0 15",
  "16 -> 2    3 32769 16",
  "16 -> 3    28673 134188470 28",
  "16 -> 4    15 32775 16",
  "16 -> 5    31745 33522654 26",
  "16 -> 6    64513 33552693 26",
  "16 -> 7    65025 16776620 25",
  "16 -> 8    255 32895 16",
  "16 -> 9    65409 4194134 23",
  "16 -> 10   65473 2097037 22",
  "16 -> 11   65505 1048544 21",
  "16 -> 12   65521 524167 20",
  "16 -> 13   65529 262136 19",
  "16 -> 14   65533 131065 18",
  "16 -> 15   1 0 1",
  "16 -> 16   1 0 0",
]
`;
